### PR TITLE
systemcontract: optimize gas cost

### LINKED ISSUE #201: GVN-05 - GAS Optimizations `[Informative]`
<p>Software optimization is the process of modifying a software system to make an aspect of it work more efficiently or use less resources. This premise must be applied to smart contracts as well, so that they execute faster or in order to save GAS.</p><p>On EVM blockchain, GAS is an execution fee which is used to compensate miners for the computational resources required to power smart contracts. If the network usage is increasing, so will the value of GAS optimization.</p><p>These are some of the requirements that must be met to reduce GAS consumption:</p><ul><li><p>Short-circuiting.</p></li><li><p>Remove redundant or dead code.</p></li><li><p>Delete unnecessary libraries.</p></li><li><p>Explicit function visibility.</p></li><li><p>Use of proper data types.</p></li><li><p>Use hard-coded constant instead of state variables.</p></li><li><p>Avoid expensive operations in a loop.</p></li><li><p>Pay special attention to arithmetical operations and comparisons.</p></li></ul>
<h4>Avoid Compound Assignment Operator</h4><p>It is recommended to avoid using compound assignment operators (for example, <code>+=</code>, <code>-=</code>, <code>*=</code>, etc.) on state variables. Compound assignment operators in Solidity require multiple steps to execute and can result in increased gas costs compared to simple assignment, like <code>var = var + x</code>.</p><p>By avoiding compound assignment operators and using simple assignment instead, it is possible to reduce gas costs, making the project more efficient.</p><h5>Source Code References</h5><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L143" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L143</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L170" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L170</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L229-L231" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L229-L231</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L281" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L281</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L419" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L419</a></p></li></ul><h4>Logic Optimization</h4><p>The <code>_computeReward</code> method initializes the local variables <code>height</code>, <code>lastGasPerVote</code> and <code>latestGasPerVote</code>. However, it immediately follows with a conditional statement. If this condition is not met, the method returns 0 directly without using the previously declared variables.</p>

image

<p>It is advisable to perform the check before declaring all the variables, in order to optimize the gas consumption of the method as much as possible in the event that the Epoch has not finished.</p><h5>Source Code References</h5><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L340-L341" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L340-L341</a></p></li></ul><p>In the <code>registerCandidate</code> and <code>exitCandidate</code> methods, the contains method of the candidateList set is used to check whether a candidate already exists before adding or removing it. By reusing the <code>boolean</code> return from the call to the <code>add</code> and <code>remove</code> methods, which returns whether it was processed correctly. Therefore, it is possible to save gas by eliminating the <code>contains</code> check.</p>

image

<h5>Source Code References</h5><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L168" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L168</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L182" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L182</a></p></li></ul><p>The <code>checkVote</code> function performs a full run over <code>voters</code> to count how many have voted. The problem is that the <code>for</code> loop continues even after the necessary number of votes has already been reached in order to consider the vote successful.</p>

image

<p>It is advisable to introduce a conditional inside the <code>for</code> loop in order to stop its execution as soon as the necessary threshold of votes has been reached. This not only reduces gas consumption but also optimizes the execution time of the function.</p><h5>Source Code References</h5><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L53" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L53</a></p></li></ul><h5>Use Constants in needVote</h5><p>It has been observed that calls to the <code>needVote</code> modifier redundantly calculate hashes with static values, resulting in excessive gas consumption during transactions.</p>

image

<p>It is recommended to previously calculate the hash values ​​of these constant values ​​within the contract.Therefore, these values ​​can be accessed without the need to recalculate them on each call to the modifier, thus reducing gas consumption and improving the overall efficiency of the contract.</p><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L55" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L55</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L67" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L67</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L79" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L79</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L55" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Policy.sol#L94</a></p></li></ul><h4>Duplicate External Contract Calls</h4><p>An inefficiency has been detected in the contract's <code>needVote</code> modifier, which makes three calls to the <code>getMinners</code> function through different methods. This practice results in redundant use of resources which triggers an increase in gas consumption.</p><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L56" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L56</a></p></li></ul><h4>Increase Operation Optimization</h4><p>During the audit it was found that it is possible to optimize all the <code>for</code> loops in the project.</p><p><br/>There are two main ways to increment a variable in solidity:</p>

```
++i will increment the value of i, and then return the incremented value.
i++ will increment the value of i, but return the original value that i held before being incremented.
```

<p>Since the return of the increment operation of the <code>for</code> loop is indifferent and discarded, both <code>i++</code> or <code>++i</code> instructions are completely valid instructions. However, the <code>++i</code> instruction has a considerably lower cost compared to incrementing a variable using <code>i++</code>, for this reason it is convenient to use <code>++i</code> for the increments of the for loops.</p><p><br/>An example of this behavior can be seen below:</p>

image

<h5>References</h5><ul><li><p><a href="https://docs.soliditylang.org/en/latest/types.html#compound-and-increment-decrement-operators" rel="noopener noreferrer nofollow" target="_blank"><u>https://docs.soliditylang.org/en/latest/types.html#compound-and-increment-decrement-operators</u></a></p></li></ul><h5>Source Code References</h5><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L22" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L22</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L37" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L37</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L48" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L48</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L50" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L50</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L141" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L141</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L377" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L377</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L386" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L386</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L315" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L315</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L398" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L398</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L401" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L401</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L419" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/Governance.sol#L419</a></p></li></ul><h4>Caching Arrays Length in Loops</h4><p>Every time an array iteration occurs in a loop, the solidity compiler will read the array's length. Thus, this is an additional <code>sload</code> operation for storage arrays (100 more extra gas for each iteration except for the first) and an additional <code>mload</code> action for memory arrays (3 additional gas for each iteration except for the first).</p>

```
for (uint i; i < array.length; ++i) {
    // ...
}
```

<p>The array length (in stack) might be cached to reduce these additional costs:</p>

```
uint length = array.length;
for (uint i; i < length; ++i) {
    // ...
}
```

<p><br/>The <code>sload</code> or <code>mload</code> action is only performed once in the example above before being replaced by the inexpensive <code>dupX</code> instruction.</p><h5>Source Code References</h5><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L22" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L22</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L37" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L37</a></p></li><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L48" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L48</a></p></li></ul><h4>Dead Code</h4><p>In programming, a part of the source code that is never used is known as dead code. The execution of this type of code consumes more GAS during deployment, something that is not necessary.</p><p>The <code>safeTransfer</code> method of the <code>TransferHelper</code> library is not used during the execution of the contract, so it would be convenient to either remove it or to use it.</p><ul><li><p><a href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovReward.sol#L7" rel="noopener noreferrer nofollow" target="_blank">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovReward.sol#L7</a></p></li></ul>

### LINKED ISSUE #205: GVN-09 - Reentrancy Pattern `[Informative]`
<p>The Reentrancy attack is a vulnerability that occurs when external contract calls can make new calls to the calling contract, before the initial execution is completed. For a function, this means that the state of the contract could change in the middle of its execution as a result of a call to an untrusted contract or the use of a low-level function with an external address.</p><p>One way to mitigate this vulnerability is to use a checks-effects-interactions pattern in the contract design. In this pattern, the contract first performs all the necessary checks to ensure that the function call is valid and then performs all the state changes before interacting with other contracts or accounts.</p>
<p>In the case of the <code>needVote</code> modifier of the <code>GovernanceVote</code> contract, votes are cleared after external calls are made, so the call can be redirected to any<em> </em>method before updating the values.</p>

image

<p>By applying the security measures described, the contract can ensure that it completes all state changes before allowing any external interactions, preventing reentrancy and other types of attacks in smart contracts.</p>

**Remediation:**
<p>It is <strong>essential to always make the state changes in the storage before making transfers or calls to external contracts</strong>, in addition to implementing the necessary measures to avoid duplicated calls or chain calls to the methods. A good practice to avoid reentrancy is to organize the code according to the following pattern:</p><ol><li><p>The first step is conducting all necessary <strong>checks</strong>.</p></li><li><p>The next step is the application of all <strong>effects.</strong></p></li><li><p>All external <strong>interactions</strong> take place in the last step.</p></li></ol>

**References:**
- https://docs.soliditylang.org/en/latest/security-considerations.html#use-the-checks-effects-interactions-pattern
- https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html

**Source references:**
<p><a target="_blank" rel="noopener noreferrer nofollow" href="https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L66-L68">https://github.com/bane-labs/go-ethereum/blob/ca856ce0ad5d20490d8e7304430f4db3e6331340/contracts/solidity/GovernanceVote.sol#L66-L68</a></p>

### LINKED ISSUE #207: GVN-11 - Absence of Unit Test `[Informative]`
<p>During the security review, we identified the absence of Unit Tests, a best practice that is not only highly recommended but has also become mandatory for projects managing large amounts of capital.</p>
<p>For the safety development of any project, at Red4Sec we consider that unitary tests are essential, and its periodical execution is a fundamental practice.</p><p>Unit tests play a crucial role for the following reasons:</p><ol><li><p><strong>Early error detection:</strong> It assists in identifying and rectifying errors at an early stage of the software development process.</p></li><li><p><strong>Enhanced code quality</strong>: By confirming that each unit operates correctly in isolation, unit tests contribute to improved overall code quality.</p></li><li><p><strong>Problem identification and resolution:</strong> It facilitates pinpointing and resolve issues, both current and future.</p></li><li><p><strong>Foundation for new features:</strong> Unit tests establish a solid foundation for the development of new features, ensuring its proper integration.</p></li><li><p><strong>Confidence in code:</strong> it increases the confidence in the code, allowing for changes and updates with the assurance that existing functionalities remain working.</p></li><li><p><strong>Simplified integration:</strong> Unit tests facilitate seamless integration and enhance the collaboration process within development teams.</p></li></ol>

### PR DESCRIPTION
**Pull request by txhsl**
Close #201, #205, #207. Done, also fixed several potential bugs of `GovernanceVote`.

### CODE REVIEW

**Review by passer-byzhang — COMMENTED** [review]
It seems like fixed except *Caching Arrays Length in Loops* and *Increase Operation Optimization*

**Comment by txhsl** [comment]
This well also close #207, since other unit tests has been already merged.

File: contracts/solidity/GovernanceVote.sol [hunk 1]
```diff
@@ -2,71 +2,87 @@
...
-    function clearVote(bytes32 methodKey) internal {
-        address[] memory voters = IGovReward(govReward).getMiners();
-        for (uint i; i < voters.length; i++) {
-            delete voteMap[methodKey][voters[i]];
+    function _clearVote(bytes32 methodKey) internal {
+        address[] memory voters = voteMap[methodKey]._keys.values();
+        uint vlength = voters.length;
+        delete voteMap[methodKey]._keys._inner._values;
+        for (uint i = 0; i < vlength; i++) {
+            delete voteMap[methodKey]._keys._inner._positions[
```
Comment (chenquanyu) [inline contracts/solidity/GovernanceVote.sol]: Why not use `remove()` of `EnumerableSet`?

Comment (txhsl) [reply contracts/solidity/GovernanceVote.sol]: `remove()` contains a rearrangement of `_values`, which is not necessary in this case and costs 4% more gas.

File: contracts/solidity/GovernanceVote.sol [hunk 2]
```diff
@@ -2,71 +2,87 @@
...
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
 
 interface IGovReward {
     function getMiners() external view returns (address[] memory);
 }
 
 abstract contract GovernanceVote {
+    using EnumerableSet for EnumerableSet.AddressSet;
+    struct AddressToBytes32Map {
+        EnumerableSet.AddressSet _keys;
```
Comment (chenquanyu) [inline contracts/solidity/GovernanceVote.sol]: Maybe we can just use array instead of `EnumerableSet` to store voters here.

Comment (txhsl) [reply contracts/solidity/GovernanceVote.sol]: I revert this part, let's create a new issue and related PR to solve it.

**Review by chenquanyu — APPROVED** [review]

**Pull request merged**